[nnc] tensorexpr for quantized/aten::upsample_nearest2d

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70193
* __->__ #70189
* #70188

Differential Revision: [D33238092](https://our.internmc.facebook.com/intern/diff/D33238092)

IvanKobzarev committed:
[nnc] tensorexpr for quantized/aten::upsample_nearest2d

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238092).

IvanKobzarev committed:
Update on "[nnc] tensorexpr for quantized/aten::upsample_nearest2d"


Differential Revision: [D33238092](https://our.internmc.facebook.com/intern/diff/D33238092)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238092).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/quantization.cpp:
```diff
@@ -721,6 +721,51 @@ Tensor computeUpsampleNearest2d(
+  unpack_dim_args(c10::fmap<DimArg>(outputShape), &dims, &args);
+  // Handle separately when scale is specified? as in 'scalar_t
+  // compute_scales_value' in UpSample.h
+  auto scale_h =
+      promoteToDtype(input_height, ScalarType::Double) / output_height;
+  auto scale_w = promoteToDtype(input_width, ScalarType::Double) / output_width;
+  // TODO: will repetetive if in idx calculation will be taken out of the loop?
+  auto compute_nearest_idx =
+      [](ExprHandle scale, ExprHandle dst_index, ExprHandle input_size) {
+        return Min::make(
+            promoteToDtype(floor(dst_index * scale), ScalarType::Long),
+            input_size - 1,
+            true);
+      };
+  auto body_func = [&](std::vector<VarHandle> axes) {
+    std::vector<ExprHandle> newAxes(axes.begin(), axes.end());
+    newAxes[2] = compute_nearest_idx(scale_h, axes[2], input_height);
+    newAxes[3] = compute_nearest_idx(scale_w, axes[3], input_width);
+    return A.load(newAxes);
+  };
+  auto e = body_func(VarVectorToVarHandleVector(args));
+  BufPtr buf = alloc<Buf>(
+      "quantize_upsample_nearest2d",
+      ExprHandleVectorToExprVector(outputShape),
+      Dtype(*outputType),
+      nullptr,
+      c10::nullopt,
+      A.node()->qscale(),
+      A.node()->qzero());
```
Similarly to the previous PR, this code would probably be slightly more elegant if we used `Compute`.

ZolotukhinM reviewed (approved):
👍

IvanKobzarev committed:
Update on "[nnc] tensorexpr for quantized/aten::upsample_nearest2d"


Differential Revision: [D33238092](https://our.internmc.facebook.com/intern/diff/D33238092)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238092).
